wdl-doc/src/docs_tree.rs

```diff
@@ -0,0 +1,336 @@
+                                p { a href=(diff_paths(node.path(), base).unwrap().to_string_lossy()) { (page.name()) } }
+                            }
+                        }
+                    } @else {
+                        p class="" { (node.name()) }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Render every page in the tree.
+    pub fn render_all(&self) {
+        let root = self.root();
+
+        for node in root.depth_first_traversal() {
+            if let Some(page) = node.page() {
+                self.write_page(page.as_ref(), node.path());
+            }
+        }
+
+        self.write_homepage();
+    }
+
+    /// Write the homepage to disk.
+    fn write_homepage(&self) {
+        let root = self.root();
+        let index_path = root.path().join("index.html");
+
+        let sidebar = self.render_sidebar_component(&index_path);
+        let content = html! {
+            div class="" {
+                h3 class="" { "Home" }
+                table class="border" {
+                    thead class="border" { tr {
+                        th class="" { "Page" }
+                    }}
+                    tbody class="border" {
+                        @for node in root.depth_first_traversal() {
+                            @if node.page().is_some() {
+                                tr class="border" {
+                                    td class="border" {
+                                        a href=(diff_paths(node.path(), root.path()).unwrap().to_str().unwrap()) {(node.name()) }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        };
+
+        let html = full_page(
+            "Home",
+            html! {
+                (sidebar)
+                (content)
+            },
+            self.stylesheet_relative_to(root.path()).as_deref(),
+        );
+        std::fs::write(index_path, html.into_string()).unwrap();
+    }
+
+    /// Write a page to disk at the designated path.
+    pub fn write_page<P: Into<PathBuf>>(&self, page: &HTMLPage, path: P) {
+        let mut path = path.into();
+
+        let content = match page.page_type() {
+            PageType::Index(doc) => {
+                path = path.join("index.html");
+                doc.render()
+            }
+            PageType::Struct(s) => s.render(),
+            PageType::Task(t) => t.render(),
+            PageType::Workflow(w) => w.render(),
+        };
+
+        let stylesheet = self.stylesheet_relative_to(path.parent().unwrap());
+        let sidebar = self.render_sidebar_component(&path);
+
+        let html = full_page(
+            page.name(),
+            html! {
+                (sidebar)
+                (content)
+            },
+            stylesheet.as_deref(),
+        );
+        std::fs::write(path, html.into_string()).unwrap();
```

Comment:
We should probably functions that interact with the file system fallible rather than panic on IO error.

Reply:
good point! I pretty heavily use `unwrap()` in this PR as a shortcut, but just went through and updated some functions to return `anyhow::Result` instead and also changed some unwraps to expects.